File: optimizers/LBFGS.py
from .Optimizer import Optimizer, np


class LBFGS(Optimizer):
    """
    Классический L-BFGS с двухцикловой рекурсией.

    В отличие от PyTorch, где step(closure) делает много внутренних итераций,
    здесь один вызов next_point() — одна итерация: обновление истории пар
    (s, y), вычисление направления d = -H·g и backtracking line search
    по условию Армихо.
    """

    s_history: list[np.ndarray]
    y_history: list[np.ndarray]
    prev_x: np.ndarray | None
    prev_gradient: np.ndarray | None

    def __init__(
        self,
        initial_x: np.ndarray,
        function,
        lr: float = 1.0,
        history_size: float = 10,
        c1: float = 0.0001,
        tau: float = 0.5,
        max_ls_steps: float = 20,
    ) -> None:
        params = dict(lr=lr, history_size=history_size, c1=c1, tau=tau, max_ls_steps=max_ls_steps)
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)

        self.update_history(gradient)
        self.prev_x = self.x
        self.prev_gradient = gradient

        # вырожденные случаи: минимум уже найден или градиент не определён
        if not np.all(np.isfinite(gradient)) or np.linalg.norm(gradient) < 1e-12:
            return self.move_next(self.x)

        direction = self.compute_direction(gradient)
        if not np.all(np.isfinite(direction)) or np.dot(gradient, direction) >= 0:
            # направление не определено или не является направлением спуска —
            # откатываемся на антиградиент
            direction = -gradient

        return self.move_next(self.line_search(direction, gradient))

    def update_history(self, gradient: np.ndarray) -> None:
        if self.prev_x is None or self.prev_gradient is None:
            return

        s = self.x - self.prev_x
        y = gradient - self.prev_gradient
        sy = np.dot(s, y)

        # защита от отрицательной кривизны: такая пара сломала бы
        # положительную определённость аппроксимации гессиана
        if not np.isfinite(sy) or sy <= 1e-10:
            return

        self.s_history.append(s)
        self.y_history.append(y)
        history_size = max(int(self.params["history_size"]), 1)
        while len(self.s_history) > history_size:
            self.s_history.pop(0)
            self.y_history.pop(0)

    def compute_direction(self, gradient: np.ndarray) -> np.ndarray:
        # пустая история — обычный градиентный спуск
        if not self.s_history:
            return -gradient

        # two-loop recursion: d = -H·g без явного хранения матрицы H
        q = gradient.copy()
        rhos = []
        alphas = []
        for s, y in zip(reversed(self.s_history), reversed(self.y_history)):
            rho = 1 / np.dot(y, s)
            alpha = rho * np.dot(s, q)
            q -= alpha * y
            rhos.append(rho)
            alphas.append(alpha)

        # начальное приближение H0 = gamma * I по последней паре
        s, y = self.s_history[-1], self.y_history[-1]
        r = np.dot(s, y) / np.dot(y, y) * q

        for s, y, rho, alpha in zip(self.s_history, self.y_history, reversed(rhos), reversed(alphas)):
            beta = rho * np.dot(y, r)
            r += (alpha - beta) * s

        return -r

    def line_search(self, direction: np.ndarray, gradient: np.ndarray) -> np.ndarray:
        # backtracking по условию Армихо: уменьшаем шаг, пока функция
        # не уменьшится хотя бы на c1 * t * (g·d)
        f0 = self.function(self.x)
        slope = np.dot(gradient, direction)

        t = self.params["lr"]
        for _ in range(max(int(self.params["max_ls_steps"]), 1)):
            next_x = self.x + t * direction
            value = self.function(next_x)
            if np.isfinite(value) and value <= f0 + self.params["c1"] * t * slope:
                return next_x
            t *= self.params["tau"]

        # подходящий шаг не найден — остаёмся на месте
        return self.x

    def reset(self) -> None:
        super().reset()
        self.s_history = []
        self.y_history = []
        self.prev_x = None
        self.prev_gradient = None

File: src/shared/lib/optimization-engine/optimizers/lbfgs.ts
import type { OptimizerDescriptor } from "./types";
import { add2, dot2, norm2, scale2, sub2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient } from "@shared/lib/optimization-engine/functions";

export const lbfgsOptimizer: OptimizerDescriptor = {
  name: "LBFGS",
  params: {
    lr: { default: 1.0, description: "начальная длина пробного шага линейного поиска вдоль квазиньютоновского направления" },
    historySize: { default: 10, description: "число хранимых пар (s, y) для двухцикловой аппроксимации обратного гессиана" },
    c1: { default: 0.0001, description: "коэффициент условия Армихо — требуемая доля предсказанного убывания функции" },
    tau: { default: 0.5, description: "множитель уменьшения шага при откате (backtracking) в линейном поиске" },
    maxLsSteps: { default: 20, description: "максимум уменьшений шага в линейном поиске; если шаг не найден — остаёмся на месте" },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let sHistory: Vec2[] = [];
    let yHistory: Vec2[] = [];
    let prevX: Vec2 | null = null;
    let prevGradient: Vec2 | null = null;

    function updateHistory(grad: Vec2): void {
      if (prevX === null || prevGradient === null) return;
      const s = sub2(x, prevX);
      const y = sub2(grad, prevGradient);
      const sy = dot2(s, y);
      // negative-curvature guard: such a pair would break positive-definiteness
      // of the inverse-Hessian approximation
      if (!Number.isFinite(sy) || sy <= 1e-10) return;
      sHistory.push(s);
      yHistory.push(y);
      const historySize = Math.max(Math.trunc(params.historySize), 1);
      while (sHistory.length > historySize) {
        sHistory.shift();
        yHistory.shift();
      }
    }

    function computeDirection(grad: Vec2): Vec2 {
      if (sHistory.length === 0) return scale2(grad, -1);

      let q = grad;
      const rhos: number[] = [];
      const alphas: number[] = [];
      for (let i = sHistory.length - 1; i >= 0; i--) {
        const s = sHistory[i];
        const y = yHistory[i];
        const rho = 1 / dot2(y, s);
        const alpha = rho * dot2(s, q);
        q = sub2(q, scale2(y, alpha));
        rhos.push(rho);
        alphas.push(alpha);
      }

      const lastS = sHistory[sHistory.length - 1];
      const lastY = yHistory[yHistory.length - 1];
      let r = scale2(q, dot2(lastS, lastY) / dot2(lastY, lastY));

      for (let i = 0; i < sHistory.length; i++) {
        const s = sHistory[i];
        const y = yHistory[i];
        const rho = rhos[rhos.length - 1 - i];
        const alpha = alphas[alphas.length - 1 - i];
        const beta = rho * dot2(y, r);
        r = add2(r, scale2(s, alpha - beta));
      }

      return scale2(r, -1);
    }

    function lineSearch(direction: Vec2, grad: Vec2): Vec2 {
      const f0 = fn(x[0], x[1]);
      const slope = dot2(grad, direction);
      let t = params.lr;
      const maxLsSteps = Math.max(Math.trunc(params.maxLsSteps), 1);
      for (let i = 0; i < maxLsSteps; i++) {
        const trialX = add2(x, scale2(direction, t));
        const value = fn(trialX[0], trialX[1]);
        if (Number.isFinite(value) && value <= f0 + params.c1 * t * slope) return trialX;
        t *= params.tau;
      }
      return x;
    }

    return {
      get x() { return x; },
      params,
      next() {
        const grad = gradient(fn, x[0], x[1]);
        updateHistory(grad);
        prevX = x;
        prevGradient = grad;

        // degenerate cases: minimum already found or gradient undefined
        if (!Number.isFinite(grad[0]) || !Number.isFinite(grad[1]) || norm2(grad) < 1e-12) {
          return { x, value: fn(x[0], x[1]) };
        }

        let direction = computeDirection(grad);
        if (!Number.isFinite(direction[0]) || !Number.isFinite(direction[1]) || dot2(grad, direction) >= 0) {
          // direction undefined or not a descent direction — fall back to steepest descent
          direction = scale2(grad, -1);
        }

        x = lineSearch(direction, grad);
        return { x, value: fn(x[0], x[1]) };
      },
      reset() {
        x = initialX;
        sHistory = [];
        yHistory = [];
        prevX = null;
        prevGradient = null;
      },
    };
  },
};
